### ISSUE TITLE
Feature request: \rotatebox

### ISSUE THREAD

**Issue by meygerjos** [Khan/KaTeX#681]
I would like to make symbols upside down.  This is done in LaTeX, using the `graphicsx` package, with `\rotatebox[origin=c]{180}{\heartsuit}` (for an upsidedown `\heartsuit`).  It would be great if KaTeX could implement `\rotatebox`, or even just the special case `\rotatedbox[origin=c]{180}`!

[The suggestion that I make this feature request](https://gitter.im/Khan/KaTeX?at=590097d38bb56c2d11e4d027)

**Comment by kevinbarabash** [comment]
I think this will be pretty easy to get working right for single glyphs or a simple horizontal run.  I worry about the center be wrong for larger layouts.

**Comment by edemaine** [comment]
This is related to #587 (horizontal metrics). To support `\rotatebox` in general, we'd need an understanding of the size of a box.  I still think this is possible, once we have horizontal metric data from fonts (#632), though it's going to be an effort to accumulate the box size.  That said, we may need these box sizes for many things (`\multicolumn` in `tabular` is on my mind) so it'd be worth trying, once #632 is in.

At the HTML level, this should be implementable via [CSS3's rotate transform](https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/rotate).  The tricky part is to get the `transform-origin` correct, which is where we need the box size.

**Comment by kevinbarabash** [comment]
If we apply the rotate transform to the <div> containing the box that we want to rotate it should work correctly for some cases because the default transform origin is 50%, 50%.  That being said, in looking at some of our layouts in the DOM inspector, sometimes there are elements that are partially outside the bounds of the visible contents of the layout which may or may not cause issues with where the origin actually appears.

**Comment by edemaine** [comment]
@kevinbarabash Hmm, definitely sounds like it's worth an experiment.  If it works in many/useful cases, probably worth including an initial version.

**Comment by kevinbarabash** [comment]
<img width="809" alt="screen shot 2017-04-30 at 11 03 54 am" src="https://cloud.githubusercontent.com/assets/1044413/25565453/c258d302-2d94-11e7-95d7-7daa1c5669b6.png">

I removed `overflow: auto` from the container that KaTeX was rendering in and added `display: inline-block; transform: rotate(310deg);`

**Comment by meygerjos** [comment]
Here is a `javascript` workaround until this is implemented in KaTeX.  Posting because it might be useful to other people who want upside-down symbols and come across this thread.

```
var scripts = document.getElementsByTagName("script");
for (var i = 0; i < scripts.length; i++) {
    var script = scripts[i];
    
    if (script.type.match(/^math\/tex/)) {
        var text = script.text === "" ? script.innerHTML : script.text;

	/* Removing comments is important because kramdown adds a
	TeX-commented CDATA wrapper to any math including the
	characters < and &. KaTeX unfortunately doesn't respect
	TeX-style comments (as MathJax does), so we need to help it
	out here.  The reason that kramdown needs to add a CDATA
	wrapper is that for some reason (idky) XHTML doesn't like '<'
	and '&' in a script tag unless there's a CDATA wrapper.  XHTML
	will see the wrapper because it doesn't respect TeX-style
	comments.  For more information see
	https://github.com/gettalong/kramdown/issues/292 and follow
	the links. */
	
	var textWithoutComments = text.replace(/%.*/g, '');

        var options = {
	    displayMode: script.type.match(/mode\s*=\s*display/),
	    macros: {
	    "\\heart": "\\heartsuit",
	    "\\butt": "\\yen" // replace "\\butt" with something that won't be used that often, like the yen symbol
	    //"\\butt": "\\rotatebox[origin=c]{180}{\\heartsuit}}"
	    }
	};
        script.insertAdjacentHTML(
	    "beforebegin",
	    katex.renderToString(textWithoutComments, options));
    }
}

// Replace yen symbol with inverted \heart
var katexs = document.getElementsByClassName("katex-html");
for (var i=0; i<katexs.length; i++) {
    var spans = katexs[i].getElementsByTagName("span");
    for (var j=0; j<spans.length; j++) {
        var span = spans[j];
        if (span.innerHTML === "¥"){
            span.innerHTML = "♡";
	    span.style = "display: inline-block; transform: rotate(180deg)";
        }
    }
}
```

**Comment by ronkok** [comment]
Instead of `\rotatebox`, maybe we should try to support `\scalebox` from the graphicx package. Then one could write: `\raisebox{\depth}{\scalebox{1}[-1]{WOW}}` and the result would be:
![wow](https://user-images.githubusercontent.com/16403058/30241936-8dbfd5d6-9541-11e7-9230-810d16858e23.png)

I think `\scalebox` would be much easier to implement than `\rotatebox`.

**Comment by ronkok** [comment]
Here's another option: That is, revive the [work](https://github.com/Khan/KaTeX/pull/119) that @kevinbarabash did to support`\class` and `\cssId`. Then the OP could set a class on a character and write a CSS `transform: scaleY(-1)` for the class.

Besides, the `\class` capability would be awesome in many ways. I think it ought to be in the KaTeX core, not an extension.

**Comment by kevinbarabash** [comment]
@ronkok why should this be part of core?

**Comment by ronkok** [comment]
Clarification: I don't mean to imply that `\class` or `\cssId` are in any LaTeX package. They are clearly non-standard, if LaTeX = standard. I just think that those two functions could add so much goodness that they should be in every KaTeX installation. One should not have to set any special rendering options to get that power.

**Comment by kevinbarabash** [comment]
@ronkok I think they'd be useful too, but some people may not want to allow that amount of freedom, especially when dealing with user input.  There should be at the very least an easy way to opt out.  Long term I think it would be cool to have ways to create custom builds, but that's a ways out.

**Comment by ronkok** [comment]
That makes too much sense to deny. I hadn't thought things all the way through.

**Comment by edemaine** [comment]
FWIW, MathJax's [HTML extension](http://docs.mathjax.org/en/latest/tex.html#html) supports `\class` and `\cssId` and `\style` (and `\href` from the `url` package).  So there's definitely reason to support them (compatibility with MathJax).  [You probably already knew this, hence the name of these macros.]  Opting out for security seems related to the recent "limit lengths" feature.

**Comment by ronkok** [comment]
In that spirit then, we could have a rendering option `omitClass`. If true, KaTeX would ignore the `\class` and `\cssId` functions.

**Comment by dainiak** [comment]
Are there any concrete plans for CSS extensions then? A lot seems to have changed in the architecture of KaTeX since 2014 effort for adding `\cssid` and `\class`. Is it at all possible now?
